### PR TITLE
Issue 48 form component

### LINKED ISSUE #48: Form Component
## Basic Information
Create a reusable form consisting of field components.

UPDATE:
Title and subtitle seen in image below can probably reuse components from #35.

image

So overall `Form` component might look like this
```html
<Form>
  <Heading text="something" subtext="optional" />
  <TextInput val=... />
   ...
</Form>
```

image

- [x] #49

- [x] #71

- **Location**: `/components/form.tsx`
- **Behaviour**: 
    - Store user input in the input fields
    - button allows progression to other pages/views
    - When required fields not filled in, button is greyed out
- **Purpose**: Used by frame 1581 login/sign up, 1552 see tasks, 1572 request extra change, 1578 new task

### PR DESCRIPTION
**Pull request by SodaVolcano**
## Summary
Created (very over-engineered) `<Form>` component consisting of a collection of input components (submit button is NOT included, add it yourself)
- ~~Pass it a JSON object defining the fields and it'll internally manage the states and titles/subtitles~~
- You now just use the input components directly inside `<Form>`
    - So we don't clutter the code with a bunch of `useState(...)`
- ~~Form data is represented as `(name, value)` pairs to the `onSubmit` function (only inputs with `name` attribute are included)~~ Is now a dictionary instead of a list
- Example usage is in `component-showcase.tsx`

Also removed margins from callouts and inputs, should be set by the user.

image

Values passed to `onSubmit(...)` (note: last field is missing `name` so it's not included in the list of form data)
image

# Related issue

- Resolve #48

### CODE REVIEW

File: client/src/components/ui/form.tsx
Comment (ErikaKK) [inline client/src/components/ui/form.tsx:1]: Should add some value validation: only when all the input values are provided can the user submit the form. 
Is it better to make a form component that looks the same as the new task form?
image

**Review by ErikaKK — CHANGES_REQUESTED** [review]

File: client/src/components/ui/form.tsx
Comment (SodaVolcano) [reply client/src/components/ui/form.tsx:1]: I fould out about `useContext()` so I'm rewriting this component so it's a bit more flexible and easier to use while still allowing it to manage the states of the inputs

**Comment by yunho7687** [comment]
Screenshot 2024-07-09 at 5 02 56 PM
there is a conflict now. Please resolve this conflict before able to merge

**Comment by SodaVolcano** [comment]
Note: used `<Button>` variant `inactive` which will be pushed by #107

**Review by yunho7687 — APPROVED** [review]

**Review by ErikaKK — APPROVED** [review]
It looks good. May create new issue for input required validation with style change in the submit button. @ArcueidShiki can have a look at it

**Pull request merged**